src/vm/comsynchronizable.cpp

```diff
@@ -1055,7 +1055,6 @@ FCIMPL1(INT32, ThreadNative::GetApartmentState, ThreadBaseObject* pThisUNSAFE)
         // ApartmentState.Unknown to our callers.
```

Comment:
This comment says that this method is intentionally initializing COM to not return ApartmentState.Unknown. It means that somebody thought about it and intentionally made this change.

Returning `ApartmentState.Unknown` here is a breaking change. If we want to make that breaking change, we need to understand the impact, why it was done this way in the first place, etc.

Reply:
@danmosemsft do you know why  "Returning ApartmentState.Unknown" was done in the first place ?

Reply:
This code existed at least until the move from Source Depot in 2010, SDV won't let me go further back.

What I observed in the issue:
> Main thread: 
> * On Core, starts off uninitialized, but GetApartmentState returns MTA (should be Unknown). Because it's uninitialized, you can successfully SetApartmentState to STA or MTA and that initializes it.
> * On Full FX, starts off as specified in attribute, else MTA. GAS returns the value accurately and because it is already  initialized you can't SAS to a different value.

>Secondary threads:
> * On both platforms, GAS accurately reports Unknown. You can set to MTA or STA to intialize it, until you have started the thread at which point GAS returns MTA unless you requested STA and SAS fails.

It looks like EnsureComStarted() here=>thread->SetApartment(Thread::AS_InMTA..)=> sets the state if it's Unknown, pends the CoInit if thread isn't started. Subsequent calls to EnsureComStarted() return true. Subsequent direct calls to SetApartment always succeed if thread isn't started.

The managed SetApartment=>thread->SetApartment with the requested value, or throws if it's already running.

When a thread starts, PrepareApartmentAndContext()=>thread->SetApartment(..) which is what ensures CoInit actually happens to whatever it was prepped with.

Secondary threads always return Unknown initially because EnsureComStarted() was already called on the main thread so g_fComStarted was set so calls to EnsureComStarted() return true.

Hmm.

Reply:
We should return the apartment state that accurately represents what the state of the thread will be when it's started. We should not change the behavior of what apartment state is chosen (aside from not ignoring the attribute anymore).

If (assuming no attributes) Core is currently CoInitializing the main thread to MTA, then we are correctly returning MTA. So this line should be reverted and you will never get Unknown apartment state on the main thread.

Reply:
The state of Main thread is Unknown when it is started (assuming no attribute) as we are able to set it to STA or MTA.

Reply:
@Anipik that's what I thought, but it's only true if you haven't called GAS already. This

```c#
            Thread t = Thread.CurrentThread;
            Console.WriteLine(t.GetApartmentState());
            t.SetApartmentState(ApartmentState.STA);
```
throws.

The reason being this call in GAS to EnsureComStarted(). So we should never return Unknown because the act of getting the value sets it to MTA (if it was Unknown) !

Reply:
fyi @tannergooding

Reply:
@danmosemsft  shouldn't we just start the COM and keep the state as Unknown (if it was Unknown) In GAS ?

Reply:
No because then the value would be different than .NET Framework even though the behavior is the same -- in both runtimes, the main thread ends up as MTA (unless there is an STA attribute or you set STA).

There is still one difference, in .NET Framework you cannot do SAS to something else on the main thread even if you never ran GAS. So this fails on .NET Framework and works on .NET Core:
```c#
        public static void Main(string[] args) // No attribute
        {       
            Thread t = Thread.CurrentThread;
            t.SetApartmentState(ApartmentState.STA);
```
Put in a GAS before the SAS and it will fail on Core also.

I don't know why that is, you would have to debug, but the key thing is what the thread is CoIinitialized to when it starts, and it's MTA in both cases, so ideally we would behave the same as .NET Framework here also.

Reply:
BTW I don't think it's partiuclarly important to mimic that SAS can't be called even without GAS.